### ISSUE TITLE
Multi-GPU 2080ti performance issues

### ISSUE THREAD

**Issue by jaymefosa** [ultralytics/yolov5#1860]
## ❔Question
Training time for 1, 2, and 4 GPUs decreases. (60, 36, 20 min respectively). However after that training time increases. At 8 GPUs training time is around 30 minutes. Does anyone have any recommendations for improving this performance?

**Comment by glenn-jocher** [comment]
@jaymefosa 👋 Hello, thank you for your interest in our work! This issue seems to lack the minimum requirements for a proper response, or is insufficiently detailed for us to help you. Please note that most technical problems are due to:

- **Your modified or out-of-date code**. If your issue is not reproducible in a new `git clone` version of this repo we can not debug it. Before going further run this code and verify your issue persists:
```bash
$ git clone https://github.com/ultralytics/yolov5 yolov5_new  # clone latest
$ cd yolov5_new
$ python detect.py  # verify detection

# CODE TO REPRODUCE YOUR ISSUE HERE
```

- **Your custom data.** If your issue is not reproducible in one of our 3 common datasets ([COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), or [VOC](https://github.com/ultralytics/yolov5/blob/master/data/voc.yaml)) we can not debug it. Visit our [Custom Training Tutorial](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) for guidelines on training your custom data. Examine `train_batch0.jpg` and `test_batch0.jpg` for a sanity check of your labels and images.

- **Your environment.** If your issue is not reproducible in one of the verified environments below we can not debug it. If you are running YOLOv5 locally, verify your environment meets all of the [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies specified below. If in doubt, download Python 3.8.0 from https://www.python.org/, create a new [venv](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/), and install requirements.

If none of these apply to you, we suggest you close this issue and raise a new one using the 🐛 **Bug Report template**, providing screenshots and **minimum viable code to reproduce your issue**. Thank you!


## Requirements

Python 3.8 or later with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including `torch>=1.7`. To install run:
```bash
$ pip install -r requirements.txt
```


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab Notebook** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Kaggle Notebook** with free GPU: [https://www.kaggle.com/models/ultralytics/yolov5](https://www.kaggle.com/models/ultralytics/yolov5)
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) 
- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)

## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.

**Comment by wudashuo** [comment]
Months ago, when I tested Multi-GPU training, I found a similar problem.
I had two GPU servers, each of them had four 1080Ti. 
When I use one machine(4 * 1080Ti) with DDP mode`python -m torch.distributed.launch --nproc_per_node 4`, the time per epoch is about 4 mins, faster than the normal mode with `--device 0,1,2,3`.
But when I use two machine(4 * 1080Ti + 4 * 1080Ti) with DDP mode, the time is 5.5 mins, slower than one machine.
According to glenn and Nano(DDP contributer), multi-GPU training isn't as efficient as single GPU training, multi-machine training isn't as efficient as single machine training.
https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training#issuecomment-681312340
https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training#issuecomment-681375139
[check this pic for training time with multi-gpu](https://user-images.githubusercontent.com/9899957/88168921-57702b00-cc45-11ea-8aab-2d020e1e45b9.png)
I don't have enough time and machine to test it, so I still can't figure out the exactly reason why 8 GPUs is slower than 4 GPUs，both in your case and in my case.

**Comment by jaymefosa** [comment]
@wudashuo Thanks for adding your experience. I have another recent datapoint to add to this thread. Using the exact same dataset, Git pull, and pip installs, on the LLambda V100 8x server, appropriate speedups are seen all the way to 8x GPU. This is using the DDP method outlined in the multi-gpu guide.

Given that the datasets were the same, SSD drive was used in both cases, Git pulls and installs were the same, the GPU model type is the most likely looking suspect. @wudashuo was using a 1080ti, this case is a 2080ti, and it seems @glenn-jocher does his testing/training for the repo on a V100?

Another minor data point to add is that the 2080ti machine has a total of 10GPUs in it, and not 8.

If anyone else has any other feedback to add that might help discover the reason for the training discrepancy it would be very much appreciated.

**Comment by glenn-jocher** [comment]
@jaymefosa I think the more-purpose built (and more expensive) GPU servers like the 8x V100 servers etc are more likely to be higher performance machines with higher core counts, faster NVMe drives, and also more likely to come with the associated inter-GPU NVLinks that Multi-GPU training really requires to achieve best performance.

I don't know what part of the slowdown may be intentional hobbling of consumer GPU servers on NVIDIA's part, how much is associated with the rest of the server hardware, what the NVLink bridges contribute to the picture, and of course the effect of the dataset itself. 

Naturally we can't speak to performance on 3rd party machines, but performance on repeatable hardware, like GCP VMs and AWS P3 instances is well benchmarked and multi-gpu performance behavior there is as expected.
